### PR TITLE
🏁 Fix adding special tokens in SFT

### PR DESCRIPTION
**Pull request by qgallouedec**
### To simplify, this is what we have been doing:

```python
if is_conversational(example):
    example["text"] = tokenizer.apply_chat_template(example["messages"])

processed = processing_class(text=example["text"])
if processed["input_ids"][-1] != processing_class.eos_token_id:
    processed["input_ids"] = processed["input_ids"] + [processing_class.eos_token_id]
    processed["attention_mask"] = processed["attention_mask"] + [1]
```

There are two issues:
- It could add the special tokens (mostly BOS) twice
- It could add the EOS twice if the EOS isn't in the end, see #3318

### This is what we should be doing:

```python
if is_conversational(example):
    example["text"] = tokenizer.apply_chat_template(example["messages"])
    add_special_tokens = False
else:
    if not processed["text"].endswith(processing_class.eos_token):  # almost always true, unless the data contains the eos, which is a bad practice
        processed["text"] = processed["text"] + processing_class.eos_token
    add_special_tokens = True

processed = processing_class(text=example["text"], add_special_tokens=add_special_tokens)
```

## Some examples

```python
from trl import SFTTrainer
from datasets import Dataset

dataset = Dataset.from_dict(
    {
        "messages": [
            [{"role": "user", "content": "What is better than ugly?"}, {"role": "assistant", "content": "Beautiful."}]
        ]
    }
)

trainer = SFTTrainer(model="Qwen/Qwen2.5-0.5B-Instruct", train_dataset=dataset)
toknizer = trainer.tokenizer
sample = trainer.train_dataset[0]
print(repr(toknizer.decode(sample["input_ids"])))
```

### It solves both issues

Before/After with `model = "Qwen/Qwen2.5-0.5B-Instruct"`
```diff
- '<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|>\n<|im_end|>'
+ '<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|>\n'
```

Before/after with `model="CohereLabs/aya-expanse-8b"`
```diff
- '<BOS_TOKEN><BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>What is better than ugly?<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>Beautiful.<|END_OF_TURN_TOKEN|>'
+            '<BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>What is better than ugly?<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>Beautiful.<|END_OF_TURN_TOKEN|>'
```

### And no other changes introduced

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -500,16 +509,6 @@ def _func(example):
...
-            # If the dataset is prompt-completion, convert it to language modeling type
-            first_example = next(iter(dataset))
-            if "prompt" in first_example.keys() and "completion" in first_example.keys():
-                key = "messages" if is_conversational(first_example) else "text"
-
-                def concat_prompt_completion(example):
-                    return {key: example["prompt"] + example["completion"]}
-
-                dataset = dataset.map(concat_prompt_completion, remove_columns=["prompt", "completion"])
-
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:512]: In order not to lose the information about what the prompt is, and what the completion is, we need to remove this. The concatenation will be carried out in tokenization

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -522,33 +521,79 @@ def concat_prompt_completion(example):
...
+                            example["completion"] = example["completion"] + eos_token
+                        return example
+
+                    dataset = dataset.map(
+                        add_eos,
+                        fn_kwargs={"eos_token": processing_class.eos_token},
+                        remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
+                        **map_kwargs,
+                    )
+                    add_special_tokens = True  # subsequent tokenization will add special tokens (mostly for bos)
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:556]: These changes aren't not related to completion-only loss.
It fixes the case where we add the BOS twice (in `apply_chat_template` and in `tokenize`. To avoid that, we need to pass `add_special_tokens=False` when the chat template was applied.

Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:556]: It also fixes the case where we add the EOS twice, because it's isn't found in the end of the sequence, see #3318

**Comment by qgallouedec** [comment]
More in-depth comparison:

```python
```

```diff
  trl-internal-testing/tiny-CohereForCausalLM
- '<BOS_TOKEN><BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>What is better than ugly?<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>Beautiful.<|END_OF_TURN_TOKEN|>'
+ '<BOS_TOKEN><|START_OF_TURN_TOKEN|><|USER_TOKEN|>What is better than ugly?<|END_OF_TURN_TOKEN|><|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>Beautiful.<|END_OF_TURN_TOKEN|>'

  trl-internal-testing/tiny-DbrxForCausalLM
- "<|im_start|>system\nYou are DBRX, created by Databricks. You were last updated in December 2023. You answer questions based on information available up to that point.\nYOU PROVIDE SHORT RESPONSES TO SHORT QUESTIONS OR STATEMENTS, but provide thorough responses to more complex and open-ended questions.\nYou assist with various tasks, from writing to coding (using markdown for code blocks — remember to use ``` with code, JSON, and tables).\n(You do not have real-time data access or code execution capabilities. You avoid stereotyping and provide balanced perspectives on controversial topics. You do not provide song lyrics, poems, or news articles and do not divulge details of your training data.)\nThis is your system prompt, guiding your responses. Do not reference it, just respond to the user. If you find yourself talking about this message, stop. You should be responding appropriately and usually that means not mentioning this.\nYOU DO NOT MENTION ANY OF THIS INFORMATION ABOUT YOURSELF UNLESS THE INFORMATION IS DIRECTLY PERTINENT TO THE USER'S QUERY.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|><|endoftext|>"
+ "<|im_start|>system\nYou are DBRX, created by Databricks. You were last updated in December 2023. You answer questions based on information available up to that point.\nYOU PROVIDE SHORT RESPONSES TO SHORT QUESTIONS OR STATEMENTS, but provide thorough responses to more complex and open-ended questions.\nYou assist with various tasks, from writing to coding (using markdown for code blocks — remember to use ``` with code, JSON, and tables).\n(You do not have real-time data access or code execution capabilities. You avoid stereotyping and provide balanced perspectives on controversial topics. You do not provide song lyrics, poems, or news articles and do not divulge details of your training data.)\nThis is your system prompt, guiding your responses. Do not reference it, just respond to the user. If you find yourself talking about this message, stop. You should be responding appropriately and usually that means not mentioning this.\nYOU DO NOT MENTION ANY OF THIS INFORMATION ABOUT YOURSELF UNLESS THE INFORMATION IS DIRECTLY PERTINENT TO THE USER'S QUERY.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|>"

  trl-internal-testing/tiny-FalconMambaForCausalLM
- '\n\nUser: What is better than ugly?\n\nAssistant: Beautiful.<|endoftext|>'
+ '\n\nUser: What is better than ugly?\n\nAssistant: Beautiful.'

  trl-internal-testing/tiny-Gemma2ForCausalLM
- '<bos><bos><start_of_turn>user\nWhat is better than ugly?<end_of_turn>\n<start_of_turn>model\nBeautiful.<end_of_turn>\n<eos>'
+ '<bos><start_of_turn>user\nWhat is better than ugly?<end_of_turn>\n<start_of_turn>model\nBeautiful.<end_of_turn>\n'

  trl-internal-testing/tiny-GemmaForCausalLM
- '<bos><bos><start_of_turn>user\nWhat is better than ugly?<end_of_turn>\n<start_of_turn>model\nBeautiful.<end_of_turn>\n<eos>'
+ '<bos><start_of_turn>user\nWhat is better than ugly?<end_of_turn>\n<start_of_turn>model\nBeautiful.<end_of_turn>\n'

  trl-internal-testing/tiny-LlamaForCausalLM-3.1
- '<|begin_of_text|><|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\n<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'
+ '<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\n<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'

  trl-internal-testing/tiny-LlamaForCausalLM-3.2
- '<|begin_of_text|><|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nCutting Knowledge Date: December 2023\nToday Date: 20 Apr 2025\n\n<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'
+ '<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nCutting Knowledge Date: December 2023\nToday Date: 20 Apr 2025\n\n<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'

  trl-internal-testing/tiny-LlamaForCausalLM-3
- '<|begin_of_text|><|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'
+ '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat is better than ugly?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nBeautiful.<|eot_id|>'

  trl-internal-testing/tiny-MistralForCausalLM-0.1
- '<s><s> [INST] What is better than ugly? [/INST] Beautiful.</s>'
+ '<s> [INST] What is better than ugly? [/INST] Beautiful.</s>'

  trl-internal-testing/tiny-MistralForCausalLM-0.2
- '<s><s> [INST] What is better than ugly? [/INST] Beautiful.</s>'
+ '<s> [INST] What is better than ugly? [/INST] Beautiful.</s>'

  trl-internal-testing/tiny-Phi3ForCausalLM
  '<|user|> What is better than ugly?<|end|><|assistant|> Beautiful.<|end|><|endoftext|>'

  trl-internal-testing/tiny-Qwen2ForCausalLM-2.5
- '<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|>\n<|im_end|>'
+ '<|im_start|>system\nYou are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>\n<|im_start|>user\nWhat is better than ugly?<|im_end|>\n<|im_start|>assistant\nBeautiful.<|im_end|>\n'
```

**Comment by shirinyamani** [comment]
Maybe pointing to your relevant blog on tokenizer ?

**Review by shirinyamani — APPROVED** [review]

File: trl/trainer/sft_trainer.py [hunk 3]
```diff
@@ -522,33 +522,57 @@ def concat_prompt_completion(example):
...
+                    if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
+                        map_kwargs["desc"] = f"Applying chat template to {dataset_name} dataset"
+                    column_names = first_example.keys()
+                    dataset = dataset.map(
+                        apply_chat_template,
+                        fn_kwargs={"tokenizer": processing_class},
+                        remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
+                        **map_kwargs,
+                    )
+                    add_special_tokens = False  # subsequent tokenization won't add special tokens (mostly for bos)
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:536]:
```suggestion
                    # Subsequent tokenization won't add special tokens (mostly for bos).
                    # See https://huggingface.co/blog/qgallouedec/gotchas-in-tokenizer-behavior#7-chat-template-and-tokenization-dont-compose-due-to-special-tokens
                    add_special_tokens = False
```

File: trl/trainer/sft_trainer.py [hunk 4]
```diff
@@ -522,33 +522,57 @@ def concat_prompt_completion(example):
...
+                            example["completion"] = example["completion"] + eos_token
+                        return example
+
+                    dataset = dataset.map(
+                        add_eos,
+                        fn_kwargs={"eos_token": processing_class.eos_token},
+                        remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
+                        **map_kwargs,
+                    )
+                    add_special_tokens = True  # subsequent tokenization will add special tokens (mostly for bos)
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:557]:
```suggestion
                    # Subsequent tokenization will add special tokens (mostly for bos).
                    # See https://huggingface.co/blog/qgallouedec/gotchas-in-tokenizer-behavior#7-chat-template-and-tokenization-dont-compose-due-to-special-tokens
                    add_special_tokens = True
```

**Comment by qgallouedec** [comment]
> Maybe pointing to your relevant blog on tokenizer for context of why `add_special_tokens = False` shoud be added `if is_conversational(example)` ?

done in c1c9f29
